src/Simplify.cpp

```diff
@@ -4246,36 +4246,95 @@ class Simplify : public IRMutator {
     }
 
     void visit(const Shuffle *op) {
-        if (op->is_interleave()) {
-            // Mutate the vectors
-            vector<Expr> new_vectors;
-            bool changed = false;
-            for (Expr vector : op->vectors) {
-                Expr new_vector = mutate(vector);
-                if (!vector.same_as(new_vector)) {
-                    changed = true;
-                }
-                new_vectors.push_back(new_vector);
+        if (op->is_extract_element() &&
+            (op->vectors[0].as<Ramp>() ||
+             op->vectors[0].as<Broadcast>())) {
+            // Extracting a single lane of a ramp or broadcast
+            if (const Ramp *r = op->vectors[0].as<Ramp>()) {
+                expr = mutate(r->base + op->indices[0]*r->stride);
+            } else if (const Broadcast *b = op->vectors[0].as<Broadcast>()) {
+                expr = mutate(b->value);
+            } else {
+                internal_error << "Unreachable";
             }
-            int terms = (int)new_vectors.size();
+            return;
+        }
 
-            // Try to collapse an interleave of broadcast into a single broadcast.
-            const Broadcast *b1 = new_vectors[0].as<Broadcast>();
-            if (b1) {
-                bool can_collapse = true;
-                for (size_t i = 1; i < new_vectors.size() && can_collapse; i++) {
-                    if (const Broadcast *b2 = new_vectors[i].as<Broadcast>()) {
-                        Expr check = mutate(b1->value - b2->value);
-                        can_collapse &= is_zero(check);
+        // Mutate the vectors
+        vector<Expr> new_vectors;
+        bool changed = false;
+        for (Expr vector : op->vectors) {
+            Expr new_vector = mutate(vector);
+            if (!vector.same_as(new_vector)) {
+                changed = true;
+            }
+            new_vectors.push_back(new_vector);
+        }
+
+        // Only try to simplify shuffles of loads that aren't concats
+        // of vectors. This is a bit of a hacky heuristic to avoid
+        // undoing the work of AlignLoads for Hexagon.
+        if (!op->is_concat() || new_vectors[0].type().is_scalar()) {
+            // Try to convert a shuffled load into a shuffle of a load.
```

Comment:
I rephrased the comment to be a bit more precise. What it's trying to do is convert a load with shuffled indices into a shuffle of a dense load. Your example is replacing a slice(concat(load, load, load)) with load(slice(concat(idx0, idx1, idx2))), which is the opposite, but also, we don't do this to concat vectors (so as not to undo the work of AlignLoads... I wish there were a more elegant way to avoid that. 

I suppose the goal of this is simplification, and arguably, dense loads are the simpler thing (maybe not from the IR perspective, but definitely so for codegen), so this heuristic isn't so unreasonable I guess.